### ISSUE TITLE
[Task]: Support more Beam portable schema types as Python types

### ISSUE THREAD

**Issue by ahmedabu98** [apache/beam#25946]
### What needs to happen?

Beam portable schemas include primitive and more complex types (represented as [logical types](https://github.com/apache/beam/tree/master/sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/logicaltypes)). Some of these types are supported in the Python SDK: https://github.com/apache/beam/blob/99202b237e364bf77f40b6da0ec22cb7b17c37d0/sdks/python/apache_beam/typehints/schemas.py#L23-L41

When necessary, Python classes are created to represent a portable type. For example, see Timestamp below:
https://github.com/apache/beam/blob/99202b237e364bf77f40b6da0ec22cb7b17c37d0/sdks/python/apache_beam/utils/timestamp.py#L45

There are some missing portable types in the Python SDK (e.g. Date, DateTime) that we should add support for to make the cross-language experience more smooth.

### Issue Priority

Priority: 2 (default / most normal work should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [ ] Component: IO connector
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

**Comment by pavleec** [comment]
JSON type is also missing in Python SDK 😕

**Comment by unography** [comment]
Hi @ahmedabu98 , currently `GEOGRAPHY` as a data type isn't supported, and it throws error here: https://github.com/apache/beam/blob/master/sdks/python/apache_beam/io/gcp/bigquery_tools.py#L110-L123

Are there plans to add support for it?

**Comment by jd185367** [comment]
I'd also like to bump this as needed for using `WriteToBigQuery` in Python:

https://github.com/apache/beam/blob/fa9eb2fe17f5f96b40275fe7b0a3981f4a52e0df/sdks/python/apache_beam/io/gcp/bigquery.py#L1887

Google recommends using the `STORAGE_WRITE_API` method in their [Dataflow Best Practices](https://cloud.google.com/dataflow/docs/guides/write-to-bigquery#best-practices-streaming), which requires passing this transform the `schema` argument for a table. But since many of our BigQuery tables have a `DATE` or `DATETIME` column, which isn't supported yet for these schemas in Python, we aren't able to use this.

> As of Beam 2.60.0, we haven't found a current workaround - e.g. specifying our `DATE` columns as `TIMESTAMP` in the Python schema seems to fail either when Beam tries to actually write to BigQuery, or at some point when the Java code is executing and doing its own conversion. If anyone knows a workaround for this, I'd appreciate it.

**Comment by liferoad** [comment]
cc @Abacn

**Comment by liferoad** [comment]
Added this to 2.66 and could be done for 2.67

**Comment by damccorm** [comment]
I'm going to bump this to 2.67.0 since it is not a release blocker. If we get to it before the release we can add it back to the milestone

**Comment by ng-oliver** [comment]
> hello - i'd also like to bump this
> 
> whenever the `schema` argument provided contains `timestamp`, `date, `datetime` the SDK will assume the data has the micro attribute which isn't the case in python native datetime
> 
> [beam/sdks/python/apache_beam/utils/timestamp.py](https://github.com/apache/beam/blob/99202b237e364bf77f40b6da0ec22cb7b17c37d0/sdks/python/apache_beam/utils/timestamp.py#L63)
> 
> Line 63 in [99202b2](/apache/beam/commit/99202b237e364bf77f40b6da0ec22cb7b17c37d0)
> 
>  self.micros = int(seconds * 1000000) + int(micros) 
> The legacy `method=STREAMING_INSERT` is 10x easier to use as it handles reading JSON and writing to BigQuery seamlessly. Appreciate an upgrade on `method=STORAGE_API` to align with Google's [pricing incentive](https://cloud.google.com/bigquery/pricing#data_ingestion_pricing) that `stroage_api` has 2TB free ingestion per month and is half price from `streaming_insert` on a pay-as-you-go basis.

i have a temporary workaround that, for each bigquery timestamp column in my json, convert it to the Beam Timestamp class.

The json would flow smoothly into WriteToBigQuery(method = 'STORAGE_WRITE_API')

```
class ParseJsonLineFn(beam.DoFn):
    def __init__(self, convert_to_beam_timestamp=False):
        self.convert_to_beam_timestamp = convert_to_beam_timestamp

    def process(self, line):
        if not line.strip():
            return  # Skip empty lines

        try:
            json_obj = json.loads(line)

            # Timestamp conversion
            if self.convert_to_beam_timestamp:
                for field in ["timestamp_1", "timestamp_2"]:
                    if field in json_obj:
                        json_obj[field] = parse_to_beam_timestamp(json_obj[field])

            yield json_obj

        except Exception as e:
            logging.error(f"Failed to parse JSON line: {e}")
            raise
```

**Comment by Abacn** [comment]
The MicroInstant issue should be fixed by #35243, in upcoming Beam 2.66.0 release

**Comment by jschroed5** [comment]
I think we're running into this with `IcebergIO` in python as well. An iceberg table with a field of type `timestamp` causes an error during pipeline setup:
```
ValueError: No logical type registered for URN 'beam:logical_type:datetime:v1'
```

**Comment by VIKTORVAV99** [comment]
We are also running into this issue as we are trying to key and group elements by datetime. It's possible to work around in several ways but it just adds more code that should not be needed.

**Comment by ahmedabu98** [comment]
I've created a doc to outline new logical types for Timestamp and DateTime, with configurable precision levels: https://s.apache.org/beam-timestamp-strategy

**Comment by ahmedabu98** [comment]
> It's possible to work around in several ways but it just adds more code that should not be needed.

@VIKTORVAV99 curious to know what approach you're taking! Are you working with Python-native [datetime](https://docs.python.org/3/library/datetime.html) type?

**Comment by VIKTORVAV99** [comment]
> > It's possible to work around in several ways but it just adds more code that should not be needed.
> 
> @VIKTORVAV99 curious to know what approach you're taking! Are you working with Python-native [datetime](https://docs.python.org/3/library/datetime.html) type?

Yes I am, currently the workaround we are using is simply converting it to a timestamp and then back to a datetime object. We lose the timezone but that's not super important for our usecase as we ensure everything is UTC beforehand.

**Comment by Amar3tto** [comment]
Should we move this to 2.69.0 ?

**Comment by unography** [comment]
@ahmedabu98 @Amar3tto would it also be possible to take a look at adding support for `GEOGRAPHY` data type?

**Comment by Abacn** [comment]
Geography type is underway: go/beampr/36121

remove milestone for now since this has been rolled over for several times and not appeared to be blocking

**Comment by PDGGK** [comment]
Hi, I've been reviewing the design doc for portable schema types (https://s.apache.org/beam-timestamp-strategy).

I'm very interested in this topic and considering it as a potential GSoC 2026 project scope. I noticed Python currently has `JdbcDateType`/`JdbcTimeType` (using `beam:logical_type:javasdk_*:v1` URNs) but lacks the standard portable URNs.

To get familiar with the codebase, I'd like to start by implementing `TimeLogicalType` with URN `beam:logical_type:time:v1` as a POC. The implementation would follow the existing pattern in schemas.py (similar to JdbcTimeType but using INT64 nanoseconds representation as per Java's Time.java).

Would this be a reasonable starting point?

I'm an existing contributor with merged PRs (#37298).

**Comment by ahmedabu98** [comment]
Hey @PDGGK, sorry I just saw your comment. I think that's a good approach! 

I'm currently working on adding a portable Date type (see #37830). It could be a good template for adding a portable type, and understanding which parts of the codebase you need to touch.

If you're still interested in this, please open a PR and tag me as a reviewer
